Implement RFC 027 File Verification

stevendanna opened the pull request:
This implements user-suppliable file content verification per RFC
027. Users can supply a block, string, or symbol to the `verify`
resource attribute.  Blocks will be called, string will be executed as
shell commands (respecting the same options as not_if and only_if), and
symbols can be used to access built-in registered validations.

tyler-ball commented on lib/chef/resource/file.rb:
```diff
@@ -115,6 +117,18 @@ def manage_symlink_source(arg=nil)
+
+      def verify(command=nil, opts={}, &block)
+        if ! (command.nil? || [String, Symbol].include?(command.class))
+          raise ArgumentError, "verify requires either a string, symbol, or a block"
```
A user could call the no-arg, no-block `verify` and it would make it past this argument validation - seems to me that shouldn't be possible.

stevendanna replied:
I'm a bit confused but I might be misunderstanding.  Without arguments or a block, what is there to validate?  In that case, this method returns the current set of validations, much like the other set_or_return style methods in this class.

tyler-ball replied:
Gah, you are right, that makes sense.  Thanks!

tyler-ball commented on lib/chef/resource/file.rb:
```diff
@@ -115,6 +117,18 @@ def manage_symlink_source(arg=nil)
+        end
+
+        if command || block_given?
+          @user_verifications << Verification.new(self, command, opts, &block)
```
I agree with @lamont-granquist  - I don't think the `user_` is any more informative than `file_` would be, so leave it off if there are no naming collisions.

stevendanna replied:
Will change on next push, thanks.

tyler-ball commented on lib/chef/resource/file/verification.rb:
```diff
@@ -0,0 +1,123 @@
+
+class Chef
+  class Resource
+    class File < Chef::Resource
```
Do you have to duplicate that `File` is a subclass of `Chef::Resource` here?  If Ruby allows it, I would think you would just want `class File` here.

lamont-granquist replied:
Pretty certain the interpreter gets pissed if you don't

stevendanna replied:
I believe this is necessary to to avoid a superclass mismatch since 'chef/resource/file/verification' is required from 'chef/resource/file'.

tyler-ball commented on lib/chef/resource/file/verification.rb:
```diff
@@ -0,0 +1,123 @@
+      # class Chef
+      #  class Resource
+      #    class File::Verification::Foo < Chef::Resource::File::Verification
+      #      provides :noop
```
The documentation above says to use the `register` method - which is correct?

stevendanna replied:
The last commit changed `register` to `provides` but it seems I missed that it was used twice in the comments.  Will fix on the next push, nice catch.

tyler-ball commented on lib/chef/resource/file/verification.rb:
```diff
@@ -0,0 +1,123 @@
+        end
+
+        def verify_block(path, opts)
+          @block.call(path)
```
`opts` is not used

stevendanna replied:
Currently, the opts hash is mostly there for future use.  I can add a comment mentioning that.

tyler-ball replied:
Meh, I don't think the comment is necessary - you can just tell me to stop being so nitpicky :smile:

tyler-ball commented on lib/chef/resource/file/verification.rb:
```diff
@@ -0,0 +1,123 @@
+                          Chef::GuardInterpreter::ResourceGuardInterpreter.new(@parent_resource, command, @command_opts)
+                        end
+          interpreter.evaluate
+        end
```
I don't like the duplication of `GuardInterpreter` in here.  If we want to re-use this logic, we should make those classes more general-purpose.  Right now they have implementation details specific to the fact that they are being ran as Guards.

There is also currently a bug around `ResourceGuardInterpreter` that means they won't get evaluated in `why_run` mode.  Granted, you wouldn't get to this point in `why_run` mode anyways.

stevendanna replied:
>  Right now they have implementation details specific to the fact that they are being ran as Guards.

Taking a look at those classes, it isn't clear to me what implementation details you are referring to.  From my reading, DefaultGuardInterpreter simply shells out and returns the result of `success?` on the command's status.  

ResourceGuardInterpreter is a bit more complex but it seems like the general goal of that class is already what we want here ("Run a command and tell me if it succeeded" with the caveat that in this special case we want to run the command in a context that makes sense given the parent resource).  Modulo bugs in that class, I can't really to determine what I would do to make it more general purpose.  

If we care about keeping an interface similar to not_if/only_if then I really think we should reuse this class in some form.  If we don't care about this, all of this could likely be replaced with:

```
     begin
        shell_out(command, @command_opts).status.success?
      rescue Chef::Exceptions::CommandTimeout
        Chef::Log.warn "Command '#{command}' timed out"
        false
     end
```

tyler-ball replied:
I agree with you that if we want to maintain the not_if/only_if interface then it is appropriate to use the guard interpreters.  Mostly I felt weird referencing a GuardInterpreter outside the context of a conditional (not_if/only_if).  But I can see how `verify` is related to these concepts.

I think my only opinion now is this code could be DRYed up by making https://github.com/opscode/chef/blob/8632c46e3ae6faa08ddfe69af5450d7f450d1b0b/lib/chef/resource/conditional.rb#L128-L134 a public, class level method and calling it in `verify_command`.

stevendanna commented:
@tyler-ball @lamont-granquist Thanks for your careful review. Apologies it took me so long to get back to this.  I've made most of the recommended changes, rebased against master, and updated the CHANGELOG and RELEASE_NOTES.  Please let me know if you have further changes you'd like to see.

lamont-granquist commented:
yep, i'm good, :shipit:

juliandunn commented:
Nice work @stevendanna and may I say how glad I am that we have appveyor integration now in additio to Travis.

tyler-ball commented:
:+1: Thanks for de-duplicating that code!  It looks like this needs to be rebased again, but other than that LGTM!

stevendanna commented:
rebased

lamont-granquist commented:
:shipit:

jdmundrawala commented:
@stevendanna this needs a rebase
